app/src/main/java/com/teamltt/carcare/DemoActivity.java

```diff
@@ -59,12 +59,91 @@ public void onReceive(Context context, Intent intent) {
     private boolean connecting = false;
     // True while the app has a connected socket to the BT device
     private boolean connected = false;
+    // UUID that is required to talk to the OBD II adapter
     private UUID uuid = new UUID(0L, 0x1101L);
 
+    // A button view that opens the socket to the OBD adapter. Text on the buttons serves as connection status
     private Button connectButton;
+
+    // Serial port input stream. Read to get responses from the OBD
     private InputStream inputStream;
+    // Serial port output stream. Write to send requests to the OBD
     private OutputStream outputStream;
 
+
+    /**
+     * This task takes control of the device's bluetooth and opens a socket to the OBD adapter.
+     * If successful, it starts another "communicate" AsyncTask
+     */
+    AsyncTask<Void, Void, Void> connectTask = new AsyncTask<Void, Void, Void>() {
+        @Override
+        protected Void doInBackground(Void... params) {
+            try {
+                // Android advises to cancel discovery before using socket.connect()
+                bluetoothAdapter.cancelDiscovery();
+                socket.connect();
```

Comment:
The trace is probably more complicated than it needs to be, but no, it will not be null here. We can talk about it if you would like. Basically, connectTask is only executed after the socket is created. The socket is created/opened when the user clicks the connect button, but there are a lot of places that it could fail (bluetooth permissions, adapter not found...)